frontend/src/scenes/web-analytics/webAnalyticsLogic.ts

```diff
@@ -20,77 +27,167 @@ interface QueryTile extends BaseTile {
 }
 
 interface TabsTile extends BaseTile {
+    activeTabId: string
+    setTabId: (id: string) => void
     tabs: {
+        id: string
         title: string
         linkText: string
         query: QuerySchema
-    }
+    }[]
 }
 
 export type WebDashboardTile = QueryTile | TabsTile
 
+export enum SourceTab {
+    REFERRING_DOMAIN = 'REFERRING_DOMAIN',
+    UTM_SOURCE = 'UTM_SOURCE',
+    UTM_CAMPAIGN = 'UTM_CAMPAIGN',
+}
+
+export enum DeviceTab {
+    BROWSER = 'BROWSER',
+    OS = 'OS',
+    DEVICE_TYPE = 'DEVICE_TYPE',
+}
+
+export enum PathTab {
+    PATH = 'PATH',
+    INITIAL_PATH = 'INITIAL_PATH',
+}
+
 export const initialWebAnalyticsFilter = [] as WebAnalyticsPropertyFilters
 
+const setOncePropertyNames = ['$initial_pathname', '$initial_referrer', '$initial_utm_source', '$initial_utm_campaign']
+const hogqlForSetOnceProperty = (key: string, value: string): string => `properties.$set_once.${key} = '${value}'`
```

Comment:
This will work for a limited subset of keys, but you can use `hogql.identifier(key)` (from `import { hogql } from './utils'`) to actually safely escape all keys.